### PR TITLE
Introduce Pretty, a Cypher pretty printer

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

File: community/cypher/frontend-3.2/src/main/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/Pretty.scala [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+
+    case Foreach(variable, collExp, updates) =>
+      "FOREACH" <+> parens(expr(variable) <+> "IN" <+> expr(collExp) <+> "|" <> nest(line <> vlist(updates.map(show))) <> line)
+
+    case Match(optional, pattern, hints, where) =>
+      val paths = pattern.patternParts.map(show).map(group).toList
+      val OPTIONAL: Doc = if (optional) "OPTIONAL" <> space else emptyDoc
+      val PATTERN = nest(group(softline <> folddoc(paths, _ <> comma <> softline <> _)))
+      val HINTS = if (hints.isEmpty) emptyDoc else line <> vlist(hints.map(show))
+      group(OPTIONAL <> "MATCH") <> PATTERN <> HINTS <> showWhere(where)
```
Comment (craigtaverner) [inline community/cypher/frontend-3.2/src/main/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/Pretty.scala]: Does group() do anything here?

File: community/cypher/frontend-3.2/src/test/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/PrettyTest.scala [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  test("should upcase multiple keywords 2") {
+    actual("match (a) where a.name='A' return a.age as SomethingTotallyDifferent") should equal(
+      expected("MATCH (a)%n  WHERE a.name = 'A'%nRETURN a.age AS SomethingTotallyDifferent")
+    )
+  }
+
+  test("should not break CREATE UNIQUE") {
+    actual("start me=node(3) match p1=(me)-[*2]-(friendOfFriend) " +
```
Comment (craigtaverner) [inline community/cypher/frontend-3.2/src/test/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/PrettyTest.scala]: Do we support `start n=node(#)` syntax in any recent version? I thought this was removed already in 2.1?

**Review by craigtaverner — COMMENTED** [review]
Would it be useful (or even possible?) to add a test that reparses the pretty printed output, and compares the AST to the first parse? That could be applied to a lot of existing cypher test code to verify correctness and completeness of the pretty printer.

File: community/cypher/frontend-3.2/src/test/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/PrettyTest.scala (hunk 2 above)
Comment (systay) [reply community/cypher/frontend-3.2/src/test/scala/org/neo4j/cypher/internal/frontend/v3_2/prettifier/PrettyTest.scala]: You are correct. The parser and AST are a little more forgiving though, to be able to give helpful error messages. I've tried supporting what the parser and AST support, and not look at what the semantic checks say.

**Comment by systay** [comment]
Let's re-open the PR when it's ready to go in

**Pull request closed without merging**